### ISSUE TITLE
`nix --version` does a surprising DNS lookup

### ISSUE THREAD

**Issue by matklad** [NixOS/nix#5441]
`nix-shell` + `direnv` became very slow for me (10+ seconds), and the culprit seemed to be two timing-outing dns requests to my router when `nix --version` is executing. Intuitively, `nix --version`shouldn't involve networking at all

```
λ nix --version
nix (Nix) 2.3.16
```

See for more details https://discourse.nixos.org/t/direnv-nix-shell-is-very-slow-after-the-last-update/15700

**Comment by cab404** [comment]
It also does this for everything with `--offline` flag.

**Comment by tomberek** [comment]
Potentially fixed by: https://github.com/NixOS/nix/pull/5384

**Comment by jorsn** [comment]
Nix looks up `this.pre-initializes.the.dns.resolvers.invalid.` and waits for the timeout.
Until the fix, a quick&dirty workaround is to map the address to localhost by putting
~~~
127.0.0.1 this.pre-initializes.the.dns.resolvers.invalid.
~~~
in `/etc/hosts`, or, in NixOS, by setting
~~~
hosts."127.0.0.1" = [ "this.pre-initializes.the.dns.resolvers.invalid." ]
~~~
in `configuration.nix`.

**Comment by baloo** [comment]
made two backports of #5384 for:
 - nix2.3 #5659 
 - nix2.4 #5658

**Comment by baloo** [comment]
> Indeed, this.pre-initializes.the.dns.resolvers.invalid is what I saw in strace!
> 
> I guess, there are two orthogonal fixes here:
> 
>     * don’t do this for —version at all
For context: this was placed here, and done early in the initialization process to avoid any thread safety problems. This was the easiest way to go.
> 
>     * use a very short timeout for DNS lookup
Sadly, this is not possible with the current nss APIs. This can only be done by modifying `/etc/resolv.conf`.

**Comment by baloo** [comment]
Fix for master, 2.3 and 2.4 have been merged. We can close this issue.

**Comment by baloo** [comment]
fix is available in nix 2.5.0 and 2.5.1
